### ISSUE TITLE
[core] RFC: arbitrary property values with XML binding

### ISSUE THREAD

**Issue by oowekyala** [pmd/pmd#973]
### The need

Atm, some rules in the code style category try to enforce a naming convention (eg a suffix) that should be respected by all the subtypes of a type. These are:
* LocalHomeNamingConvention
* LocalInterfaceSessionNamingConvention
* MDBAndSessionBeanNamingConvention
* RemoteInterfaceNamingConvention
* RemoteSessionInterfaceNamingConvention

1. I think they are too specific for PMD
2. They are all implemented the same way, and it could be nice to abstract that away. PMD itself uses similar conventions of its own, eg the name of all interfaces extending Node should end with "Node"

It would be nice to have one rule to rule them all, which would be configurable. Unfortunately there's no easy way to do that with our current properties.

My initial idea, was to introduce a StringMapProperty, which would have Map<String, String> as its value. We could then mention the conventions we'd like to follow using eg 
```java
<property name="typeToRegex" value="n.s.pmd.Node -> .+Node" />
```

This wouldn't scale to using many conventions, and is fragile and ugly because it depends of the tokenization of a single string value into a full map. Plus, the rule-side API exposes only strings so you have to do the conversions by hand, and there's no type checking/ validity checking at parse time   

Moreover, for each convention, we might want to specify whether it applies eg only for concrete classes, only for interfaces, etc. So we'd need a nicer way to enter all that information cleanly in the xml. 

### One possible answer

My current idea to achieve that is to introduce a new type of property, which can take any object as a value, and whose values you can set using an XML element, binding XML attributes to bean properties. For example (I didn't focus on the schema, we probably should use some other element that `value`):
```java
<property name="conventions" >
  <value forType="n.s.pmd.lang.ast.Node" regexConvention=".+Node" onInterfaces="true">
  <value forType="n.s.pmd.lang.metrics.Metric" regexConvention=".+Metric" onlyConcreteClasses="true">
</property>
```
Here each `value` element creates a new value for the property "conventions", assuming it's a multi-valued property. Each of these values is an object which has setters for properties "forType", "regexConvention", "onInterfaces", "onlyConcreteClasses", and default values for some of these. The rule then receives a list of those objects to work on via the property.

This kind of flexible property would also help many rules be more powerful and expressive, for example, the *NamingConventionsRules could, instead of just having a regex for each convention, allow setting also the minimum and maximum name length, thus superseding ShortClassName, ShortMethodName,  ShortVariable and LongVariable with more flexibility. (I assume here that #972 is implemented, since this text was extracted from that issue)

Any ideas how to implement that? I think jaxb might do something similar but I have no first-hand experience with it

**Comment by oowekyala** [comment]
Actually I think we may not need that

Current rule equality is defined by rule name + language:

http://github.com/pmd/pmd/blob/7a6bc498264ad87f3783e6dfde0b7e4bc9d4bcbf/pmd-core/src/main/java/net/sourceforge/pmd/RuleSet.java#L196

So we may have an "abstract" rule with eg the following XPath, modulo some other configuration properties:

```ruby
//ClassOrInterfaceDeclaration[ExtendsList/ClassOrInterfaceType[pmd-java:typeIs($superType) and not(matches(@Image, $regexConvention))]
```

And a user could "instantiate" it several times in the same ruleset by referencing it, overriding the name and setting the properties. The only loose point here is that the properties can have no sensible default value, so the best behaviour would be to fail if they're not set IMO. This feature is lacking from the properties framework for now.

Also, this definition of rule equality is *not* documented as of yet, and would have to become public API if we decide this solution is ok

**Comment by jsotuyod** [comment]
@oowekyala yes, I think that solution is ok. There are already several places where PMD assumes rule names to be unique, so exposing that and enforcing it internally seems adequate. Also, other tools such as Checkstyle use similar approaches base don rule names.

**Comment by oowekyala** [comment]
@jsotuyod Thanks for your input! I'll probably start working on that once I'm done with #972

**Comment by oowekyala** [comment]
Closed in favour of #1432
